File: core/examples/micro_bench_ops.js
// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
// deno-lint-ignore-file no-empty

const NR_OP_CALLS = 100000;

function assert(cond) {
  if (!cond) {
    throw Error("assert");
  }
}

// Bench functions
function binBenchSync(opName) {
  for (let i = 0; i < NR_OP_CALLS; i++) {
    Deno.core.binOpSync(opName, 42);
  }
}

function binBenchSyncErr(opName) {
  for (let i = 0; i < NR_OP_CALLS; i++) {
    try {
      Deno.core.binOpSync(opName, 42);
    } catch {}
  }
}

async function binBenchAsync(opName, binSize) {
  assert(NR_OP_CALLS % binSize === 0);
  for (let i = 0; i < NR_OP_CALLS / binSize; i++) {
    const promises = [];
    for (let i = 0; i < binSize; i++) {
      promises.push(Deno.core.binOpAsync(opName, 42));
    }
    await Promise.allSettled(promises);
  }
}

function jsonBenchSync(opName) {
  for (let i = 0; i < NR_OP_CALLS; i++) {
    Deno.core.jsonOpSync(opName, 42);
  }
}

function jsonBenchSyncErr(opName) {
  for (let i = 0; i < NR_OP_CALLS; i++) {
    try {
      Deno.core.jsonOpSync(opName, 42);
    } catch {}
  }
}

async function jsonBenchAsync(opName, binSize) {
  assert(NR_OP_CALLS % binSize === 0);
  for (let i = 0; i < NR_OP_CALLS / binSize; i++) {
    const promises = [];
    for (let i = 0; i < binSize; i++) {
      promises.push(Deno.core.jsonOpAsync(opName, 42));
    }
    await Promise.allSettled(promises);
  }
}

// Time function
const nowBytes = new Uint8Array(8);
function opNow() {
  Deno.core.binOpSync("op_now", 0, nowBytes);
  return new DataView(nowBytes.buffer).getFloat64();
}

function formatNumber(nr) {
  return nr.toFixed(2).replace(/\d(?=(\d{3})+\.)/g, '$&,');  // 12,345.67
}

async function main() {
  Deno.core.ops();
  Deno.core.registerErrorClass("Error", Error);

  // deno-lint-ignore no-undef
  const shouldPrint = getShouldPrint();

  const benches = {
    "bin_sync": binBenchSync,
    "bin_sync_err": binBenchSyncErr,
    "bin_async": binBenchAsync,
    "bin_async_err": binBenchAsync,
    "json_sync": jsonBenchSync,
    "json_sync_err": jsonBenchSyncErr,
    "json_async": jsonBenchAsync,
    "json_async_err": jsonBenchAsync,
  }

  let results = {}
  // Sync bench
  for (const type of ["bin", "json"]) {
    for (const error of [false, true]) {
      const name = `${type}_sync${error && "_err" || ""}`;
      const start = opNow();
      benches[name](name);
      const end = opNow();
      results[name] = end - start;
    }
  }

  // Async bench
  for (const type of ["bin", "json"]) {
    for (const error of [false, true]) {
      for (const bin of [1, 10, 100, 1000]) {
        const name = `${type}_async${error && "_err" || ""}`;
        const start = opNow();
        await benches[name](name, bin);
        const end = opNow();
        results[`${name}_${bin}`] = end - start;
      }
    }
  }

  const opsPerSecResults = {};
  const microsecPerOpResults = {};
  for (let bench in results) {
    const opsPerSec = (NR_OP_CALLS * 1000) / results[bench];
    const microsecPerOp = 1000 * results[bench] / NR_OP_CALLS;

    if (shouldPrint) Deno.core.print(`${bench}: ${formatNumber(opsPerSec)} op/s - ${formatNumber(microsecPerOp)} µs/op\n`);

    opsPerSecResults[bench] = opsPerSec;
    microsecPerOpResults[bench] = microsecPerOp;
  }

  if (!shouldPrint) {
    Deno.core.print(JSON.stringify({
      "ops_per_sec": opsPerSecResults,
      "microsec_per_op": microsecPerOpResults,
    }));
  }
}

main();

File: core/examples/micro_bench_ops.rs
// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.

#[macro_use]
extern crate log;

use deno_core::error::custom_error;
use deno_core::error::AnyError;
use deno_core::JsRuntime;
use deno_core::OpState;
use std::env;
use std::time::Instant;

pub type StartTime = Instant;

struct Logger;

impl log::Log for Logger {
  fn enabled(&self, metadata: &log::Metadata) -> bool {
    metadata.level() <= log::max_level()
  }

  fn log(&self, record: &log::Record) {
    if self.enabled(record.metadata()) {
      println!("{} - {}", record.level(), record.args());
    }
  }

  fn flush(&self) {}
}

fn create_js_runtime() -> JsRuntime {
  let mut runtime = JsRuntime::new(Default::default());
  {
    let op_state = runtime.op_state();
    let mut state = op_state.borrow_mut();
    state.put::<StartTime>(StartTime::now());
  }

  // Fixed size sync msgs
  runtime.register_op(
    "json_sync",
    deno_core::json_op_sync(|_state, arg: u32, _buf| {
      debug!("json_sync");
      Ok(arg)
    }),
  );
  runtime.register_op(
    "json_sync_err",
    deno_core::json_op_sync(|_state, _arg: u32, _buf| -> Result<u32, AnyError> {
      debug!("json_sync_err");
      Err(custom_error(
        "MicroBenchError",
        "This error is used for testing the ops performance in case of an error.",
      ))
    }),
  );
  runtime.register_op(
    "bin_sync",
    deno_core::bin_op_sync(|_state, arg: u32, _buf| {
      debug!("bin_sync");
      Ok(arg)
    }),
  );
  runtime.register_op(
    "bin_sync_err",
    deno_core::bin_op_sync(|_state, _arg: u32, _buf| -> Result<u32, AnyError> {
      debug!("bin_sync_err");
      Err(custom_error(
        "MicroBenchError",
        "This error is used for testing the ops performance in case of an error.",
      ))
    }),
  );

  // Fixed size async msgs
  runtime.register_op(
    "json_async",
    deno_core::json_op_async(|_state, arg: u32, _buf| async move {
      debug!("json_async");
      Ok(arg)
    }),
  );
  runtime.register_op(
    "json_async_err",
    deno_core::json_op_async(|_state, _arg: u32, _buf| async {
      debug!("json_async_err");
      Err(custom_error(
        "MicroBenchError",
        "This error is used for testing the ops performance in case of an error.",
      )) as Result<u32, AnyError>
    }),
  );
  runtime.register_op(
    "bin_async",
    deno_core::bin_op_async(|_state, arg: u32, _buf| async move {
      debug!("bin_async");
      Ok(arg)
    }),
  );
  runtime.register_op(
    "bin_async_err",
    deno_core::bin_op_async(|_state, _arg: u32, _buf| async {
      debug!("bin_async_err");
      Err(custom_error(
        "MicroBenchError",
        "This error is used for testing the ops performance in case of an error.",
      )) as Result<u32, AnyError>
    }),
  );

  runtime.register_op("op_now", deno_core::bin_op_sync(|op_state: &mut OpState, _arg: u32, buf| {
    let start_time = op_state.borrow::<StartTime>();
    let seconds = start_time.elapsed().as_secs();
    let subsec_nanos = start_time.elapsed().subsec_nanos() as f64;

    let result = (seconds * 1_000) as f64 + (subsec_nanos / 1_000_000.0);

    (&mut buf.unwrap()).copy_from_slice(&result.to_be_bytes());

    Ok(0)
  }));

  runtime
}

fn main() {
  log::set_logger(&Logger).unwrap();
  log::set_max_level(
    env::args()
      .find(|a| a == "-D")
      .map(|_| log::LevelFilter::Debug)
      .unwrap_or(log::LevelFilter::Warn),
  );

  // NOTE: `--help` arg will display V8 help and exit
  deno_core::v8_set_flags(env::args().collect());

  let mut js_runtime = create_js_runtime();
  let runtime = tokio::runtime::Builder::new_current_thread()
    .enable_all()
    .build()
    .unwrap();

  let silent = env::args()
    .find(|a| a == "-s")
    .map(|_| true)
    .unwrap_or(false);

  js_runtime
    .execute(
      "envset.js",
      if silent {
        "function getShouldPrint() { return false; }"
      } else {
        "function getShouldPrint() { return true; }"
      },
    )
    .unwrap();

  runtime.block_on(js_runtime.run_event_loop()).unwrap();

  js_runtime
    .execute("micro_bench_ops.js", include_str!("micro_bench_ops.js"))
    .unwrap();

  runtime.block_on(js_runtime.run_event_loop()).unwrap();
}
